app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -140,6 +140,9 @@ class HomeFragment : Fragment() {
         }
 
         override fun onCollectionRenamed(tabCollection: TabCollection, title: String) {
+            lifecycleScope.launch(Main) {
+                view?.sessionControlRecyclerView?.adapter?.notifyDataSetChanged()
```

Comment:
Instead of calling `notifyDataSetChanged` directly on the adapter can we dispatch a collections change? `homeFragmentStore.dispatch(HomeFragmentAction.CollectionsChange(it))`

Reply:
Yes, dispatching a collections change is how the adapter is currently notified of the change, but via the observer installed on the tab collection storage here:
https://github.com/mozilla-mobile/fenix/blob/7d5c199e5188f1d8d553a4e732fbf9b649fd38d3/app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt#L847-L850
rather than in onCollectionRenamed - both of those methods get called (on different threads) from the TabCollectionStorage code when a rename occurs.
The problem is, once we're in either of those methods, the TabCollection object has already been renamed. And since the collection that was renamed in TabCollectionStorage is the same (same object) as the collection that is currently sitting on the adapter list, when it got renamed in TabCollectionStorage it also got renamed on the adapter. So now calling the dispatch which ultimately pushes the new list of TabCollections to the adapter doesn't make the RecyclerView think it needs to update the UI, because the list currently sitting on its adapter already matches the updated collections list (data-wise (but no longer object-wise (but the new-objects list that was dispatched to the adapter is now setup for the same thing to happen again on the next rename, because it's the same - object-wise - as the list that will be sent to the rename fragment and from there to TabCollectionStorage))).

Reply:
I hadn't really thought of it explicitly before, but it's possible the notifyDataSetChanged() call I added and the existing dispatch linked above are racy.  Luckily it doesn't matter because, as I mentioned, the new dispatched list already matches the list currently on the adapter datawise, so either way, the changed collection name will appear whenever the notifyDataSetChanged() arrives.